platforms/software/dependency-management/src/main/java/org/gradle/internal/component/ResolutionCandidateAssessor.java

```diff
@@ -0,0 +1,189 @@
+import org.gradle.api.internal.attributes.AttributeValue;
+import org.gradle.api.internal.attributes.ImmutableAttributes;
+import org.gradle.internal.Cast;
+import org.gradle.internal.component.model.AttributeMatcher;
+
+import javax.annotation.Nullable;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * A utility class used by {@link ResolutionFailureHandler} to assess and classify
+ * how the attributes of candidate variants during a single attempt at dependency resolution
+ * align with the requested attributes.
+ */
+/* package */ final class ResolutionCandidateAssessor {
+    private final ImmutableAttributes requestedAttributes;
+    private final AttributeMatcher attributeMatcher;
+
+    public ResolutionCandidateAssessor(AttributeContainerInternal requestedAttributes, AttributeMatcher attributeMatcher) {
+        this.requestedAttributes = requestedAttributes.asImmutable();
+        this.attributeMatcher = attributeMatcher;
+    }
+
+    public ImmutableAttributes getRequestedAttributes() {
+        return requestedAttributes;
+    }
+
+    public AssessedCandidate assessCandidate(
+        String candidateName,
+        ImmutableAttributes candidateAttributes
+    ) {
+        Set<String> alreadyAssessed = new HashSet<>(candidateAttributes.keySet().size());
+        ImmutableList.Builder<AssessedAttribute<?>> compatible = ImmutableList.builder();
+        ImmutableList.Builder<AssessedAttribute<?>> incompatible = ImmutableList.builder();
+        ImmutableList.Builder<AssessedAttribute<?>> onlyOnConsumer = ImmutableList.builder();
+        ImmutableList.Builder<AssessedAttribute<?>> onlyOnProducer = ImmutableList.builder();
+
+        Sets.union(requestedAttributes.getAttributes().keySet(), candidateAttributes.getAttributes().keySet()).stream()
+            .sorted(Comparator.comparing(Attribute::getName))
+            .forEach(attribute -> classifyAttribute(requestedAttributes, candidateAttributes, attributeMatcher, attribute, alreadyAssessed, compatible, incompatible, onlyOnConsumer, onlyOnProducer));
+
+        return new AssessedCandidate(candidateName, candidateAttributes, compatible.build(), incompatible.build(), onlyOnConsumer.build(), onlyOnProducer.build());
+    }
+
+    private void classifyAttribute(ImmutableAttributes requestedAttributes, ImmutableAttributes candidateAttributes, AttributeMatcher attributeMatcher,
+                                   Attribute<?> attribute, Set<String> alreadyAssessed,
+                                   ImmutableList.Builder<AssessedAttribute<?>> compatible, ImmutableList.Builder<AssessedAttribute<?>> incompatible,
+                                   ImmutableList.Builder<AssessedAttribute<?>> onlyOnConsumer, ImmutableList.Builder<AssessedAttribute<?>> onlyOnProducer) {
+        if (alreadyAssessed.add(attribute.getName())) {
+            Attribute<Object> untyped = Cast.uncheckedCast(attribute);
```

Comment:
I know this was existing code, but all of these unchecked casts and coercing scare me. I have to imagine there is a better way to do this. Not saying that needs to be done now, just an observation

Reply:
Yeah, they're ugly for sure.  Yea, `Attribute`s!